### PR TITLE
Add utility functionality for trapping memory accesses

### PR DESCRIPTION
**Pull request by vekterli**
@baldersheim please review
@havardpe FYI

At its core this is a way to wrap a range of memory in a way where all non-zero writes are eventually detected and, if supported by the underlying OS and hardware, instantly detects both reads and writes to as much as possible of the underlying memory range. Implemented using `mprotect` of page-aligned regions under the hood.

### CODE REVIEW

**Review by havardpe — COMMENTED** [review]
possible improvements:
* check protected area before checking unprotected are to prefer worse error case
* protect with read and check zeroes on init to avoid before-protected race

File: vespalib/src/vespa/vespalib/util/memory_trap.cpp [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+#include <vespa/log/log.h>
+LOG_SETUP(".vespalib.util.memory_trap");
+
+using namespace std::string_view_literals;
+
+namespace vespalib {
+
+namespace {
+
+#ifdef VESPA_HAS_MPROTECT
```
Comment (baldersheim) [inline vespalib/src/vespa/vespalib/util/memory_trap.cpp:29]: I am not a fan of #ifdefs. Is it possible to gather all 5 into just a single one. That makes it far more readable and maintainable.
You get clean method definitions. One dummy set and one for mprotect.

**Review by baldersheim — CHANGES_REQUESTED** [review]
Prefer a single ifdef.
I also think that @havardpe or @toregge should take a look. I have a better eye for the gory details.

**Comment by vekterli** [comment]
> Prefer a single ifdef.

Reduced from 5 to 2. I don't want to mix includes in with the code body section, nor do I want to include Linux-specific headers on other OSes. PTAL

**Comment by vekterli** [comment]
> possible improvements:
> 
> * check protected area before checking unprotected are to prefer worse error case
> * protect with read and check zeroes on init to avoid before-protected race

Will consider these in a follow-up

**Comment by vekterli** [comment]
Actually, the relevant headers appear to be POSIX, so I can remove the `ifdef` around those, causing much happiness across the land

File: vespalib/src/vespa/vespalib/util/memory_trap.cpp (hunk 1 above)
Comment (vekterli) [reply vespalib/src/vespa/vespalib/util/memory_trap.cpp]: Fixed

**Review by baldersheim — APPROVED** [review]

**Pull request merged**